kernel/core/datamanagement/memory.cpp

```diff
@@ -3748,94 +3748,162 @@ std::vector<mu::value_type> Memory::getIndex(size_t col, const std::vector<mu::v
     return vIndex;
 }
 
-
 /////////////////////////////////////////////////
 /// \brief Calculates the simples form of a ANOVA
 /// F test
 ///
-/// \param colCategories size_t
+/// \param colCategories VectorIndex&
 /// \param colValues size_t
 /// \param _vIndex const VectorIndex&
 /// \param significance double
 /// \return AnovaResult
 ///
 /////////////////////////////////////////////////
-AnovaResult Memory::getOneWayAnova(size_t colCategories, size_t colValues, const VectorIndex& _vIndex, double significance) const
+std::vector<AnovaResult> Memory::getAnova(const VectorIndex& colCategories, size_t colValues, const VectorIndex& _vIndex, double significance) const
 {
     // Get indices
-    if (colCategories > memArray.size()
-        || !memArray[colCategories]
-        || memArray[colCategories]->m_type != TableColumn::TYPE_CATEGORICAL
+    if (colCategories[0] > memArray.size()
+        || !memArray[colCategories[0]]
+        || memArray[colCategories[0]]->m_type != TableColumn::TYPE_CATEGORICAL
         || significance >= 1.0
-        || significance <= 0.0)
+        || significance <= 0.0
+        || colCategories.size() > 2)
     {
         AnovaResult res;
         res.m_FRatio = NAN;
-        return res;
+        return std::vector<AnovaResult> {res};
     }
 
-    _vIndex.setOpenEndIndex(getElemsInColumn(colCategories)-1);
-    Memory _mem(2);
-    _mem.memArray[0].reset(memArray[colCategories]->copy(_vIndex));
-    _mem.memArray[1].reset(memArray[colValues]->copy(_vIndex));
+    bool isTwoWay = colCategories.size() == 2;
+    _vIndex.setOpenEndIndex(getElemsInColumn(colCategories[0])-1);
+
+    Memory _mem(colCategories.size()+1);
+    _mem.memArray[0].reset(memArray[colValues]->copy(_vIndex));
+    for(int i = 0; i < colCategories.size(); i++)
+        _mem.memArray[i+1].reset(memArray[colCategories[i]]->copy(_vIndex));
+
+    std::vector<std::vector<std::string>> factors;
+    for(int i = 0; i < colCategories.size(); i++)
+        factors.push_back(static_cast<CategoricalColumn*>(_mem.memArray[i+1].get())->getCategories());
 
-    const std::vector<std::string>& vCategories = static_cast<CategoricalColumn*>(_mem.memArray[0].get())->getCategories();
+    std::vector<Memory> groupedValues(std::pow(2, factors.size()) - 1);
 
-    // Copy into different columns
-    for (const auto& cat : vCategories)
+    // Copy the gropues into different columns (g1, g2, g1Xg2)
+    // For 3+ - way anova here some algo for clustering in all subsets of combinations of groups would be needed

```

Comment:
Sounds a bit like recursion could be an applicable approach? In the end, a 3-way is a combination of 3 two-way and 3 one-way ANOVAs with their complete effect. Can't that be solved like this, i.e. calling the whole thing recursive until we reach the two-way ANOVA and then unwinding until we are at the n-th-way ANOVA again? Is not required for this issue but will be interesting in the future.

Reply:
Good Point! This loop for creating all "groups" can be solved either recursive, or recursive via a bitmask.
Then the Sum of Squares for the Intersections always need the previous calculated SS of the single Factors. That also sounds like its greatly solvable via recursion. Sounds interesting! I would like to implement that later on :)

Reply:
It's true that recursion has bad performance compared to iteration, if an iteration is possible. In situations like tree traversals or similar situations, where the problem itself is recursive and can be solved without the (explicit) knowledge of the upper or lower levels, recursions are good.

Recursions also have the problem that they can provoke a stack overflow, especially if the call tree is already nested quite deeply.